From 52f9ba922f34fcd5ee30a6c62e1c8e60f1cfd29d Mon Sep 17 00:00:00 2001
From: Steve Ellcey <sellcey@caviumnetworks.com>
Date: Mon, 6 Mar 2017 14:04:28 -0800
Subject: [PATCH 2/2] Enable ILP32 mode in gdb on aarch64.

Signed-off-by: Steve Ellcey <sellcey@caviumnetworks.com>

[Patch v2 2/2] Enable ILP32 mode in gdb on aarch64
Here is the second of two gdb patches to support aarch64 ILP32 mode.
This one just has the bfd changes. Along with changes to cpu-aarch64.c
and elfnn-aarch64.c parts before, this modification to elfxx-
aarch64.c to fix gdb support in ILP32 mode.

	* elfxx-aarch64.c (_bfd_aarch64_elf_grok_prstatus): Handle
	size and offset of ILP32 executables.
	(_bfd_aarch64_elf_grok_psinfo):Ditto.

V2 changes:
	* elfxx-aarch64.c (_bfd_aarch64_elf_grok_prstatus): Correct
	the offset for ilp32 from 32 to 24.	

Signed-off-by: Manjunath Matti <mmatti@mvista.com>

Upstream-Status: Submitted [https://sourceware.org/ml/binutils/2017-03/msg00075.html]

---
 bfd/elfxx-aarch64.c | 30 +++++++++++++++++++++++++-----
 1 file changed, 25 insertions(+), 5 deletions(-)

diff --git a/bfd/elfxx-aarch64.c b/bfd/elfxx-aarch64.c
index 10ee5fa99e..932fbdbc6b 100644
--- a/bfd/elfxx-aarch64.c
+++ b/bfd/elfxx-aarch64.c
@@ -561,14 +561,25 @@ _bfd_aarch64_elf_grok_prstatus (bfd *abfd, Elf_Internal_Note *note)
       default:
 	return FALSE;
 
-      case 392:		/* sizeof(struct elf_prstatus) on Linux/arm64.  */
+      case 352:	/* sizeof(struct elf_prstatus) on Linux/aarch64 ilp32.  */
 	/* pr_cursig */
-	elf_tdata (abfd)->core->signal
-	  = bfd_get_16 (abfd, note->descdata + 12);
+	elf_tdata (abfd)->core->signal = bfd_get_16 (abfd, note->descdata + 12);
 
 	/* pr_pid */
-	elf_tdata (abfd)->core->lwpid
-	  = bfd_get_32 (abfd, note->descdata + 32);
+	elf_tdata (abfd)->core->lwpid = bfd_get_32 (abfd, note->descdata + 24);
+
+	/* pr_reg */
+	offset = 72;
+	size = 272;
+
+	break;
+
+      case 392:	/* sizeof(struct elf_prstatus) on Linux/aarch64.  */
+	/* pr_cursig */
+	elf_tdata (abfd)->core->signal = bfd_get_16 (abfd, note->descdata + 12);
+
+	/* pr_pid */
+	elf_tdata (abfd)->core->lwpid = bfd_get_32 (abfd, note->descdata + 32);
 
 	/* pr_reg */
 	offset = 112;
@@ -590,12 +601,21 @@ _bfd_aarch64_elf_grok_psinfo (bfd *abfd, Elf_Internal_Note *note)
     default:
       return FALSE;
 
+    case 124:         /* sizeof(struct elf_prpsinfo) on Linux/aarch64 ilp32.  */
+      elf_tdata (abfd)->core->pid = bfd_get_32 (abfd, note->descdata + 12);
+      elf_tdata (abfd)->core->program
+	= _bfd_elfcore_strndup (abfd, note->descdata + 28, 16);
+      elf_tdata (abfd)->core->command
+	= _bfd_elfcore_strndup (abfd, note->descdata + 44, 80);
+      break;
+
     case 136:        /* This is sizeof(struct elf_prpsinfo) on Linux/aarch64.  */
       elf_tdata (abfd)->core->pid = bfd_get_32 (abfd, note->descdata + 24);
       elf_tdata (abfd)->core->program
 	= _bfd_elfcore_strndup (abfd, note->descdata + 40, 16);
       elf_tdata (abfd)->core->command
 	= _bfd_elfcore_strndup (abfd, note->descdata + 56, 80);
+      break;
     }
 
   /* Note that for some reason, a spurious space is tacked
-- 
2.11.1

